### DISCUSSION TITLE
Fee Validation on Market orders

### DISCUSSION THREAD

**Discussion by jafri** [rodrigo-brito/ninjabot#46] [Q&A] [answered]
Trying to understand why taker/maker commission is added to validation for amounts when doing a market sell or buy on Binance

Lets say I have 10 ETH as balance, if I try to sell 10 ETH with bot currently, validate will fail ErrInsufficientFunds since it adds `TakerCommission`

How should I calculate the `size` to pass to a sell amount? Should I calculate it as 10 / (1 + b.userInfo.TakerCommission)?

**Comment by rodrigo-brito** [comment]
Hi @jafri, to sell all the open positions, you have to consider the fee. For Binance, it is 0.1%.
I normally use small sizes than 100% (risk control issues).

I recommend you to keep a small amount in (USDT, BNB, BUSD) to take these fees.
For example, when you use: `assetPosition, quotePosition, err := broker.Position(df.Pair)` for buy 100% of your quote. Use `quotePosition * 0.99` (1% reserved for fees). Then, you have a large margin of error when executing it in production.

Using `quotePosition * 0.999` (0.1% for fees) is very dangerous because the price can change in a second and the fee increases.

**Comment by jafri** [comment]
@rodrigo-brito wouldn't that make it not possible to sell 100% of your holdings?

When using Binance UI directly, I can simply select 100% of the asset to sell and it will sell 100% and automatically deduct fees

**Comment by rodrigo-brito** [comment]
Yes, but it considers the calculation of the fee in this 100% and notices that it is not accurate. After selling 100%, always rest a small amount of coins in your wallet.

Another important point, the exchange accepts precision for the order size. You can check these limits here: https://api.binance.com/api/v1/exchangeInfo

For example, this is the restrictions for BTCUSDT
```json
{
          "filterType": "PRICE_FILTER",
          "minPrice": "0.01000000",
          "maxPrice": "1000000.00000000",
          "tickSize": "0.01000000"
        },
        {
          "filterType": "PERCENT_PRICE",
          "multiplierUp": "5",
          "multiplierDown": "0.2",
          "avgPriceMins": 5
        },
        {
          "filterType": "LOT_SIZE",
          "minQty": "0.00000100",
          "maxQty": "9000.00000000",
          "stepSize": "0.00000100"
        },
        {
          "filterType": "MIN_NOTIONAL",
          "minNotional": "10.00000000",
          "applyToMarket": true,
          "avgPriceMins": 5
        },
        {
          "filterType": "ICEBERG_PARTS",
          "limit": 10
        },
        {
          "filterType": "MARKET_LOT_SIZE",
          "minQty": "0.00000000",
          "maxQty": "138.57632811",
          "stepSize": "0.00000000"
        },
        {
          "filterType": "MAX_NUM_ORDERS",
          "maxNumOrders": 200
        },
        {
          "filterType": "MAX_NUM_ALGO_ORDERS",
          "maxNumAlgoOrders": 5
        }
```

**Comment by jafri** [comment]
I understand that a little bit of coins are left due to precision size

Here is the code I am discussing

https://github.com/rodrigo-brito/ninjabot/blob/63062e8529d5b8a41cfc4cbe71d56ccc589c4991/pkg/exchange/binance.go#L135-L139

In ninjabot, if my balance is 1 BTC and I try selling 1 BTC, it will fail at validate since it tries to add taker fee on top of the 1 BTC instead of from it.

In binance API, if my balance is 1 BTC, and I pass `quantity` as 1 on a SELL MARKET order through binance, wouldn't that execute successfully?

Also, does `b.formatQuantity(symbol, quantity)` round down always, or will it also round up if the last digit is >= 5?

**Comment by rodrigo-brito** [comment]
I will make some tests, maybe it works.
About the precision, it rounds up. 

We use FloatFormat:
https://github.com/rodrigo-brito/ninjabot/blob/6b4b9b8857b47714c5bbd49a945ade01a6226037/pkg/exchange/binance.go#L258-L264

Example: https://play.golang.org/p/GG-xUA8-CNC

**Comment by rodrigo-brito** [comment]
Hi @jafri, I make some tests and are you shure, we do not need this validation. I will refactor the code and will remove it.

**Comment by jafri** [comment]
I'm not 100% sure that just removing validation is what is needed @rodrigo-brito, I just know my order with amount was failing through this library which succeeded doing the same amount in UI

**Comment by rodrigo-brito** [comment]
I will try to remove it and change the format function to avoid problems with aproximation

**Comment by rodrigo-brito** [comment, marked as answer]
Hi @jafri, i fixed this in the last release.
